### ISSUE TITLE
Question: webp support

### ISSUE THREAD

**Issue by BeebBenjamin** [openseadragon/openseadragon#2453]
We use OpenSeaDragon via Mirador.  We need to use webp to get decent data rentention for our normal maps for a threejs overlay.  I noticed there is a function to extend `FILEFORMATS`.  But it doesn't look like it's possible to do this through Mirador.

How does one use: `setImageFormatsSupported`?

**Comment by iangilman** [comment]
Does Mirador give you access to the OpenSeadragon viewer? If so, you should be able to just call:

`viewer.setImageFormatsSupported({ webp: true });`

**Comment by BeebBenjamin** [comment]
I think there is possibly a bug.  I can access other functions from the viewer, but when I try to access `setImageFormatsSupported` I get ```json
viewer.setImageFormatsSupported is not a function

**Comment by BeebBenjamin** [comment]
I think it's the `mirrador.min.js` that is missing the update.  So not a bug with OpenSeadragon.

**Comment by iangilman** [comment]
Oh, sorry... Looks like I gave you the wrong information... It's `OpenSeadragon.setImageFormatsSupported`, not `viewer.setImageFormatsSupported`. Can you get access to the base OpenSeadragon from Mirador? You shouldn't have to modify the OSD code.

**Comment by BeebBenjamin** [comment]
Not that I know of, in Mirador, viewer is the instance of OSD.  I don't think the function will work for us without adding the ability to set it via config, or exposing the function to viewer.  I have asked in IIIF slack if anyone knows how to wrangle that.  I was wondering, why OSD blocks a bunch of image types using `FILEFORMATS` anyway.  Is it to do with browser compatibility?

**Comment by BeebBenjamin** [comment]
I am making use of `preferredFormats` and for some reason OSD blocks anything other than `jpeg` or `png`.

```javascript
if (data.preferredFormats) {
                for (var f = 0; f < data.preferredFormats.length; f++ ) {
                    if ( OpenSeadragon.imageFormatSupported(data.preferredFormats[f]) ) {
                        data.tileFormat = data.preferredFormats[f];
                        break;
                    }
                }
            }
```

**Comment by iangilman** [comment]
> Not that I know of, in Mirador, viewer is the instance of OSD.

We don't want an instance of OSD, we want the OSD static namespace. Have you tried asking in the discussions or issues in https://github.com/ProjectMirador/mirador?

> I was wondering, why OSD blocks a bunch of image types using FILEFORMATS anyway. Is it to do with browser compatibility?

Correct. By default we want to only support images that work in all of OSD's supported browsers. That way you don't end up with a nasty gotcha because you only tested on some browsers. We provide the `setImageFormatsSupported` so you can support additional formats if you know what you're doing. It's generally quite straightforward, but evidently we didn't anticipate this scenario where you don't have access to the OSD namespace.

My hope is that it is possible via Mirador somehow... This isn't the only reason why it would be good for Mirador users to have access to the OSD namespace. If it's not possible, perhaps Mirador or OSD needs to be updated to support access to the namespace.

But also, now that you bring it up, webp should go on our supported formats list now that we've (recently) dropped support for IE11! That's a super simple change of course... Let me know if you want to make a PR for that.

We are a long way from making another release, though, since a number of big things are in the works here. I'm not sure what the shortest path to getting you unblocked is... Waiting for OSD5 and then for Mirador to officially adopt it is probably not a good idea.

**Comment by iangilman** [comment]
I've filed https://github.com/openseadragon/openseadragon/issues/2454 for adding webp support.

**Comment by pearcetm** [comment]
> It's generally quite straightforward, but evidently we didn't anticipate this scenario where you don't have access to the OSD namespace.

@iangilman I've come across this before - not in mirador specifically, but elsewhere where a `viewer` is exposed but `OpenSeadragon` itself isn't.

It seems like it would be pretty straightforward to add OSD as a property on `viewer` i.e. so it can be accessed by `viewer.OpenSeadragon`. Would you envision any unintended consequences of this?

**Comment by BeebBenjamin** [comment]
@iangilman Thanks for filing for webp support.  Happy to put the change in myself.  I have a workaround of building our own OSD, so if it's on master that's all that would matter in the short term.  When I do the next release of our plug-in would probably require all the changes to be available via NPM because building our own OSD is obviously not what you want for a Mirador plug-in.

@pearcetm I'll check in with the MIrador devs about accessing OSD through the name space.

**Comment by iangilman** [comment]
Yeah, being able to access the OSD namespace through the viewer is probably a good workaround. I don't think it would cause problems, but we need to make sure we document it to be clear what its purpose is. I wonder if there's a better name for it than just `viewer.OpenSeadragon`? Probably not.

Of course it won't fix all situations... I imagine there are some cases where you would need to call the OSD function before the viewer exists. When you're in a system like Mirador, you don't really have control over the life cycle of the viewer, though.

Another thought: we should probably remove the reference to OSD in `viewer.destroy` just in case there's any chance of circular references causing garbage collection issues.
